packages/xo-server/src/index.js

```diff
@@ -140,6 +148,54 @@ async function setUpPassport(express, xo) {
     res.redirect('/')
   })
 
+  express.get('/signin/otp', (req, res, next) => {
+    const { user } = req.session
+
+    if (!user) {
+      return res.redirect('/signin')
+    }
+
+    res.send(
+      signInPage({
+        error: req.flash('error')[0],
+        otp: true,
+        strategies,
+      })
+    )
+  })
+
+  express.post('/signin/otp', (req, res, next) => {
+    const { body, session } = req
+    const { user } = session
+
+    if (!user) {
+      return res.redirect('/signin')
+    }
+
+    const isValid = authenticator.check(body.otp, user?.preferences?.otp)
+    if (isValid) {
+      setToken(req, res, next)
+    } else {
+      req.flash('error', 'Invalid code')
+      return res.redirect('/signin/otp')
```

Comment:
Why do a redirect instead of directly displaying the page? (there are good reasons but I want you to be aware of them)

Reply:
To implement the PRG pattern ? :smile:

Reply:
On the first try!

![](https://media.giphy.com/media/3geedWNEOAcHC/giphy.gif)